standards.md

```diff
@@ -1,47 +1,126 @@
+* Should explain thoroughly how the new feature works
 
-- What is the problem being solved?
-- Why is this the best approach?
-- What other approaches did you consider?
-- What side effects will this approach have?
-- What future work remains to be done?
+## Tests
 
-## Coding standards
+* New features (and often/whenever possible bug fixes) have one or both of:
+  * Examples (i.e. yaml tests)
+  * End to end tests
+  * Reconciler tests
+* New types have:
+  * Validation + validation tests
+* Unit tests:
+  * Coverage should remain the same or increase
+  * Test exported functions only, in isolation:
+    * Each exported function should have tests
+    * Each test should only test one function at a time
+    * If you find yourself wanting to test an unexported function, consider whether
+      it would make sense to move the test into another package and export it
+* Test code
+  * When using cmp.Diff the argument order is always (want, got) and in the error message include (-want +got)
+    (and/or use a lib like [PrintWantGot](https://github.com/tektoncd/pipeline/blob/master/test/diff/print.go))
+  * Table driven tests: do not use the same test for success and fail cases if the logic is different
+    (e.g. do not have two sets of test logic, gated with `wantErr`)
 
-### Go
+## API Design
+
+* Avoid kubernetes specific feature in our APIs as much as possible
+* If we don’t definitely need the feature, don’t add it
+  * Must be backed up with [CI/CD specific use cases](https://github.com/tektoncd/community/blob/master/roadmap.md#mission-and-vision)
+* Prefer keeping the API simple:
+  * If you can avoid adding a new feature, don’t add it
+  * Prefer reusing existing features to adding new ones
+
+## Functionality
+
+* It should be safe to cut a release at any time, i.e. merging this PR should not
+  put us into an unreleasable state
+    * When incrementally adding new features, this may mean that a release could contain
+      a partial feature, i.e. the type specification only but no functionality
+
+## Content
+
+* Whenever logic is added that uses an image that wasn’t used before, the image used should
+  be configurable on the command line so that distributors can build images that meet their
+  support and licensing requirements
+* Refactoring should be merged separately from bug fixes and features
+  * i.e. if you refactor as part of implementing something, commit it and merge it before merging the change
+* Prefer small pull requests; if you can think of a way to break up the pull request into multiple, do it
+
+## Code
+* Reviewers are expected to understand the changes well enough that they would feel confident
+  saying the understand what is changing and why:
+  * Read through all the code changes
+  * Read through linked issues and pull requests, including the discussions
+* Reconciler changes:
+  * Avoid adding functions to the reconciler struct
+  * Avoid adding functions to the reconciler package
```

Comment:
These two are definitely pipeline-specific, and sort of look like a general smell with the reconciler package :)

Reply:
we now have at least 3 projects with reconciler (pipelines, triggers, chains - i think?) and i believe the operator is about to have a reconciler as well, so i feel like it's worth calling them out

>  sort of look like a general smell with the reconciler package :)

do you mean the underlying packages we're using? id say this is a general smell with how ppl tend to work with existing code, preferring to add more to packages and structs that already exist vs creating new ones. in particular since the reconciler is at the heart of so many of our changes, folks tend to apply this approach to the reconciler